src/openrct2/localisation/Localisation.cpp

```diff
@@ -1227,38 +1227,71 @@ money32 string_to_money(char * string_to_monetise)
 {
     const char* decimal_char = language_get_string(STR_LOCALE_DECIMAL_POINT);
     char * text_ptr = string_to_monetise;
-    int i, j, sign;
+    sint32 sign;
+    uint32 numNumbers = 0;
+    bool hasMinus = false;
+    bool hasDecSep = false;
+
     // Remove everything except numbers decimal, and minus sign(s)
-    for (i = 0; text_ptr[i] != '\0'; ++i) {
+    for (uint32 i = 0, j; text_ptr[i] != '\0'; ++i)
+    {
+        if (text_ptr[i] >= '0' && text_ptr[i] <= '9')
+        {
+            numNumbers++;
+        }
+        if (text_ptr[i] == decimal_char[0])
+        {
+            if (hasDecSep)
+                return MONEY32_UNDEFINED;
+            else
+                hasDecSep = true;
+        }
+        if (text_ptr[i] == '-')
+        {
+            if (hasMinus)
+                return MONEY32_UNDEFINED;
+            else
+                hasMinus = true;
+        }
+
         while (!(
             (text_ptr[i] >= '0' && text_ptr[i] <= '9') ||
             (text_ptr[i] == decimal_char[0]) ||
             (text_ptr[i] == '-') ||
-            (text_ptr[i] == '\0')
-        )) {
+            (text_ptr[i] == '\0')))
+        {
             // Move everything over to the left by one
-            for (j = i; text_ptr[j] != '\0'; ++j) {
+            for (j = i; text_ptr[j] != '\0'; ++j)
+            {
                 text_ptr[j] = text_ptr[j + 1];
             }
             text_ptr[j] = '\0';
         }
     }
 
-    // If first character of shortened string is a minus, consider number negative
-    if (text_ptr[0] == '-') {
-        sign = -1;
-    }
-    else {
-        sign = 1;
-    }
+    if (numNumbers == 0)
+        return MONEY32_UNDEFINED;
 
-    // Now minus signs can be removed from string
-    for (i = 0; text_ptr[i] != '\0'; ++i) {
-        if (text_ptr[i] == '-') {
-            for (j = i; text_ptr[j] != '\0'; ++j) {
-                text_ptr[j] = text_ptr[j + 1];
+    sign = 1;
+    // Check for minus signs, which are only valid at position 0.
+    for (uint32 i = 0, j; text_ptr[i] != '\0'; ++i)
```

Comment:
I think you can do this in constant time. Since you've already established there is at most one minus sign, it has to be at position 0, if at all.